storage/s3/s3.go

```diff
@@ -44,39 +45,61 @@ func New[T ingest.Identifiable](bucket string, prefix string, metafilesPrefix st
 	}
 }
 
-func (ms *minioStorage[T]) Store(ctx context.Context, element T, download func(context.Context, T) (ingest.Object, error)) (*url.URL, error) {
-	s3Key := path.Join(ms.prefix, element.ID())
-
+func (ms *minioStorage[T]) Stat(ctx context.Context, element T) (*storage.ObjectInfo, error) {
 	synced, done, err := ms.isObjectSynced(ctx, element.ID(), ms.useDone)
 	if err != nil {
 		return nil, err
 	}
 
 	if !synced {
-		object, err := download(ctx, element)
-		if err != nil {
-			return nil, fmt.Errorf("failed to get message %s: %w", element.ID(), err)
-		}
+		return nil, fs.ErrNotExist
+	}
 
-		if _, err := ms.mc.PutObject(
-			ctx,
-			ms.bucket,
-			s3Key,
-			object,
-			object.Len(),
-			minio.PutObjectOptions{ContentType: object.MimeType()}, // I guess we can remove the mime type detection because we always use tar.gz files.
-		); err != nil {
-			return nil, err
+	// If the file exists but the done file does not,
+	// let's patch this up.
+	if !done && ms.useDone {
+		if _, err := ms.mc.PutObject(ctx, ms.bucket, path.Join(ms.metafilesPrefix, doneKey(element.ID())), bytes.NewReader(make([]byte, 0)), 0, minio.PutObjectOptions{ContentType: "text/plain"}); err != nil {
+			return nil, fmt.Errorf("failed to create missing meta object for existing file: %w", err)
 		}
 	}
 
-	if !done && ms.useDone {
+	return &storage.ObjectInfo{URI: ms.url(element).String()}, nil
+}
+
+func (ms *minioStorage[T]) Store(ctx context.Context, element T, download func(context.Context, T) (ingest.Object, error)) (*url.URL, error) {
+	u := ms.url(element)
+
+	object, err := download(ctx, element)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get message %s: %w", element.ID(), err)
+	}
+
+	if _, err := ms.mc.PutObject(
+		ctx,
+		ms.bucket,
+		u.Path,
+		object,
+		object.Len(),
+		minio.PutObjectOptions{ContentType: object.MimeType()}, // I guess we can remove the mime type detection because we always use tar.gz files.
+	); err != nil {
+		return nil, err
+	}
+
+	if ms.useDone {
 		if _, err := ms.mc.PutObject(ctx, ms.bucket, path.Join(ms.metafilesPrefix, doneKey(element.ID())), bytes.NewReader(make([]byte, 0)), 0, minio.PutObjectOptions{ContentType: "text/plain"}); err != nil {
-			return nil, err
+			return nil, fmt.Errorf("failed to create matching meta object for uploaded file: %w", err)
 		}
 	}
 
-	return &url.URL{Scheme: "s3", Host: ms.bucket, Path: s3Key}, nil
+	return u, nil
+}
+
+func (ms *minioStorage[T]) url(element T) *url.URL {
+	return &url.URL{
+		Scheme: "s3",
+		Host:   ms.bucket,
```

Comment:
is this a common practice to have url like s3://bucket/key without a host?

Reply:
the bucket _is_ the host in there. If you look at URI's in the s3 console for example, they are formatted like this. the S3 endpoint confusingly isn't the host in the URL.

Reply:
ok i thought i saw s3://bucket.host/key but maybe it was https://bucket.host/key, since s3 is not aware that the actual host is not always amazon.com

Reply:
ah hmm need to take another look. it's possible. however we specify the endpoint in a flag so the pods know what to deal with. I guess the problem would be in cases where the dequeuer doesn't know what endpoint to use. but we don't currently support this; the endpoint is set in stone at startup. maybe we want to refactor in the future to make this flexible somehow, though we would need to provide credentials for ad-hoc endpoints